plugins/README.adoc

```diff
@@ -7,3 +7,256 @@ NOTE: Work in Progress...
+if not valid:
+    raise AnsibleFilterError(errors)
+return ipaddr(**updated_data)
+----
+====
+
+== All plugin types should have documentation
+[%collapsible]
+====
+Explanations::
+All plugins, regardless of type, should have documentation that describes the input parameters, outputs, and practical examples of how to use it.
+====
+
+== Use sphinx (reST) formatted docstrings
+[%collapsible]
+====
+Explanations::
+All classes, private and public functions should have a sphinx (reST) formatted docstring including all parameters, yields, raises, or returns.
+
+Rationale::
+PEP-257 states that all modules should normally have docstrings, and all functions and classes exported by a module should also have docstrings. Public methods (including the __init__ constructor) should also have docstrings. A package may be documented in the module docstring of the __init__.py file in the package directory.
+
+Examples::
+[source,python]
+----
+"""[Summary]
+
+:param [ParamName]: [ParamDescription], defaults to [DefaultParamVal]
+:type [ParamName]: [ParamType](, optional)
+...
+:raises [ErrorType]: [ErrorDescription]
+...
+:return: [ReturnDescription]
+:rtype: [ReturnType]
+"""
+----
+
+====
+
+== If Python 3 or greater, all code should have type hints and be validated with mypy
+[%collapsible]
+====
+Explanations:: Use Python type hints to document variable types.
+
+Rationale::  Type hints communicate what type a variable can be expected to be in the code. They can be consumed by static analysis tools to ensure that variable usage is consistent within the code base.
+MyPy is a static type checker/
+
+Examples::
+[source,python]
+----
+def greeting(name: str) -> str:
+    return 'Hello ' + name
+----
+
+====
+
+== The use of unittest is discouraged, use pytest instead.
+[%collapsible]
+====
+Explanations:: Use pytest for writing unit tests for plugins
+
+Rationale:: Pytest is the testing framework used by Ansible Engineering and will provide the best experience for plugin developers
+
+Examples::
+[source,python]
+----
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+import pytest
+
+from ansible.modules.copy import AnsibleModuleError, split_pre_existing_dir
+from ansible.module_utils.basic import AnsibleModule
+
+ONE_DIR_DATA = (('dir1',
+                 ('.', ['dir1']),
+                 ('dir1', []),
+                 ),
+                ('dir1/',
+                 ('.', ['dir1']),
+                 ('dir1', []),
+                 ),
+                ) 
+
+@pytest.mark.parametrize('directory, expected', ((d[0], d[2]) for d in ONE_DIR_DATA))
+def test_split_pre_existing_dir_one_level_exists(directory, expected, mocker):
+    mocker.patch('os.path.exists', side_effect=[True, False, False])
+    split_pre_existing_dir(directory) == expected
+
+----
+
+====
+
+== A complex plugin argspec should be derived from its documentation runtime if the plugin’s source is manually maintained.
```

Comment:
As we can't provide documentation or user-accessible examples I'm removing this recommendation.  It can always be proposed again in the future if the circumstance changes.